test/test_filesystem_helper_std.cpp

```diff
@@ -0,0 +1,520 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <fstream>
+#include <string>
+
+#include "rcpputils/filesystem_helper.hpp"
+#include "rcpputils/env.hpp"
+
+#ifdef _WIN32
+static constexpr const bool is_win32 = true;
+#else
+static constexpr const bool is_win32 = false;
+#endif
+
+using path = std::filesystem::path;
+
+std::string build_extension_path()
+{
+  return is_win32 ? R"(C:\foo\bar\baz.yml)" : "/bar/foo/baz.yml";
+}
+
+std::string build_double_extension_path()
+{
+  return is_win32 ? R"(C:\bar\baz.bar.yml)" : "/foo/baz.bar.yml";
+}
+
+std::string build_no_extension_path()
+{
+  return is_win32 ? R"(.\test_folder)" : R"(./test_folder)";
+}
+
+std::string build_directory_path()
+{
+  return is_win32 ? R"(.\test_folder)" : R"(./test_folder)";
+}
+
+TEST(TestFilesystemHelper, join_path)
+{
+  {
+    auto p = path("foo") / path("bar");
+    if (is_win32) {
+      EXPECT_EQ("foo\\bar", p.string());
+    } else {
+      EXPECT_EQ("foo/bar", p.string());
+    }
+  }
+
+  if (is_win32) {
+    auto p = path("foo") / path("C:\\bar");
+    EXPECT_EQ("C:\\bar", p.string());
+  } else {
+    auto p = path("foo") / path("/bar");
+    EXPECT_EQ("/bar", p.string());
+  }
+
+  {
+    // Just expect these join operations to be allowed with const paths
+    const auto p1 = path("foo");
+    auto p1_baz = p1 / "baz";
+
+    const auto p2 = path("bar");
+    auto p1_2 = p1 / p2;
+  }
+}
+
+TEST(TestFilesystemHelper, parent_path)
+{
+  {
+    auto p = path("my") / path("path");
+    EXPECT_EQ(p.parent_path().string(), path("my").string());
+  }
+  {
+    auto p = path("foo");
+    EXPECT_EQ(p.parent_path().string(), "");
+  }
+  {
+    if (is_win32) {
+      {
+        auto p = path("C:\\foo");
+        EXPECT_EQ(p.parent_path().string(), "C:\\");
+      }
+      {
+        auto p = path("\\foo");
+        EXPECT_EQ(p.parent_path().string(), "\\");
+      }
+    } else {
+      auto p = path("/foo");
+      EXPECT_EQ(p.parent_path().string(), "/");
+    }
+  }
+  {
+    if (is_win32) {
+      {
+        auto p = path("C:\\");
+        EXPECT_EQ(p.parent_path().string(), "C:\\");
+      }
+      {
+        auto p = path("\\");
+        EXPECT_EQ(p.parent_path().string(), "\\");
+      }
+    } else {
+      auto p = path("/");
+      EXPECT_EQ(p.parent_path().string(), "/");
+    }
+  }
```

Comment:
The extra scope here doesn't seem necessary, and I think we can combine these into a shorter set of `if` statements:

```suggestion
  if (is_win32) {
    {
      auto p = path("C:\\foo");
      EXPECT_EQ(p.parent_path().string(), "C:\\");
    }
    {
      auto p = path("\\foo");
      EXPECT_EQ(p.parent_path().string(), "\\");
    }
    {
      auto p = path("C:\\");
      EXPECT_EQ(p.parent_path().string(), "C:\\");
    }
    {
      auto p = path("\\");
      EXPECT_EQ(p.parent_path().string(), "\\");
    }
  } else {
    {
      auto p = path("/foo");
      EXPECT_EQ(p.parent_path().string(), "/");
    }
    {
      auto p = path("/");
      EXPECT_EQ(p.parent_path().string(), "/");
    }
  }
```

Reply:
https://github.com/ros2/rcpputils/pull/196/commits/5467b2afee010a87b82aa3701c88bf2a31efcce0